middleware/middleware.go

```diff
@@ -0,0 +1,75 @@
+package middleware
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/services/stats"
+)
+
+func MaxConcurrentRequestsMiddleware(maxConcurrentRequests int) func(http.Handler) http.Handler {
+
+	var requests = make(chan struct{}, maxConcurrentRequests)
+	return func(next http.Handler) http.Handler {
+		//this is to make sure that we don't have more than `maxClient` in-memory at any point of time. As, having more http client than `maxClient`
+		// may lead to gateway OOM kill.
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if maxConcurrentRequests != 0 {
+				select {
+				case requests <- struct{}{}:
+					defer func() {
+						<-requests
+					}()
+				default:
+					http.Error(w, http.StatusText(http.StatusServiceUnavailable), http.StatusServiceUnavailable)
```

Comment:
What does it happen if we keep returning 503 for too long? Would this affect some load balancing mechanism in the middle? Also, is this monitored somehow?

Reply:
so essentially, with this PR we want to make sure that gateway doesn't get OOM killed, if DB write is slow for some reason in comparison to ingress. So, ideally we shouldn't return too many 503's for too long. Only for sometime, when DB is under pressure, or if someone sends too many requests in a short period of time, & I believe rejecting in this case is fine. 
But, we are planning to keep a count of 503's returned & maybe add an alert for the same, if too many 503 are returned in a separate PR.
Also, I don't think this will affect load balancing mechanism in the middle. Though, not sure here. 

@atzoum @lvrach @chandumlg any thoughts?

Reply:
capturing statistics of response codes per request path (information typically available in an access log) can be part of another pull request.
Alerts on top of such metrics can follow at a latter phase

Reply:
Given the nature of this PR (rejecting requests) and given that it is a new behaviour we are introducing I'd suggest adding the metrics now. Also considering how easy you added a gauge a few lines below with `stats.DefaultStats`. Whatever the outcome whenever something like this is introduced I think it's beneficial to be able to see immediately the effects of such change. It is also not clear to @saurav-malani whether this might affect load balancing of something else in the middle so if we end up with an incident, having someone respond and see a big bunch of 503s could help them with the incident resolution.

Reply:
Unless of course we are already collecting response codes statistics in some other layer (at Netdata we were). @lvrach ?

Reply:
@fracasula we will selectively enable the limit on specific customer(s), by default it is disabled. The existing metric, number of active connections should be enough for monitoring & troubleshooting at this stage.